platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+	odp_buffer_t buf = (odp_buffer_t)(uintptr_t)ml_compl;
+	odp_event_t ev = odp_buffer_to_event(buf);
+
+	if (odp_unlikely(ml_compl == ODP_ML_COMPL_INVALID)) {
+		_ODP_ERR("Given ML completion event is invalid\n");
+		return -2;
+	}
+
+	if (odp_event_types(ev, &subtype) != ODP_EVENT_ML_COMPL ||
+	    subtype != ODP_EVENT_ML_COMPL_RUN) {
+		_ODP_ERR("Given completion event has wrong event type or subtype\n");
+		return -2;
+	}
+
+	run_result = odp_buffer_addr(buf);
+	if (result)
+		*result = *run_result;
+
+	return run_result->error_code ? -1 : 0;
+}
+
+int odp_ml_compl_load_result(odp_ml_compl_t ml_compl, odp_ml_load_result_t *result)
+{
+	odp_event_subtype_t subtype;
+	odp_ml_load_result_t *load_result;
+	odp_buffer_t buf = (odp_buffer_t)(uintptr_t)ml_compl;
+	odp_event_t ev = odp_buffer_to_event(buf);
+
+	if (odp_unlikely(ml_compl == ODP_ML_COMPL_INVALID)) {
+		_ODP_ERR("Given ML completion event is invalid\n");
+		return -2;
+	}
+
+	if (odp_event_types(ev, &subtype) != ODP_EVENT_ML_COMPL ||
+	    subtype != ODP_EVENT_ML_COMPL_LOAD) {
+		_ODP_ERR("Given completion event has wrong event type or subtype\n");
+		return -2;
+	}
+
+	load_result = odp_buffer_addr(buf);
+	if (result)
+		*result = *load_result;
+
+	return load_result->error_code ? -1 : 0;
+}
+
+void *odp_ml_compl_user_area(odp_ml_compl_t ml_compl)
+{
+	return odp_buffer_user_area((odp_buffer_t)(uintptr_t)ml_compl);
+}
+
+odp_ml_compl_t odp_ml_compl_from_event(odp_event_t event)
+{
+	_ODP_ASSERT(_odp_event_hdr_field(event, int8_t, event_type) == ODP_EVENT_ML_COMPL);
+
+	return (odp_ml_compl_t)(uintptr_t)event;
+}
+
+odp_event_t odp_ml_compl_to_event(odp_ml_compl_t ml_compl)
+{
+	return (odp_event_t)(uintptr_t)ml_compl;
+}
+
+uint64_t odp_ml_compl_to_u64(odp_ml_compl_t ml_compl)
+{
+	return (uint64_t)(uintptr_t)ml_compl;
+}
+
+void odp_ml_compl_param_init(odp_ml_compl_param_t *compl_param)
+{
+	memset(compl_param, 0, sizeof(odp_ml_compl_param_t));
+
+	compl_param->queue	= ODP_QUEUE_INVALID;
+	compl_param->event	= ODP_EVENT_INVALID;
+}
+
+int odp_ml_model_load(odp_ml_model_t model, odp_ml_load_result_t *result)
+{
+	odp_ml_load_result_t result_local;
+	int ret = -1;
+	ml_model_t *mdl = ml_model_from_handle(model);
+
+	memset(&result_local, 0, sizeof(result_local));
+
+	if (odp_unlikely(!(_odp_ml_glb->ml_config.load_mode_mask & ODP_ML_COMPL_MODE_SYNC))) {
+		_ODP_ERR("Synchronous load is not enabled!\n");
+		result_local.error_code = ML_FEATURE_NOT_SUPPORTED;
+		goto load_fail;
+	}
+
+	if (odp_unlikely(model == ODP_ML_MODEL_INVALID)) {
+		_ODP_ERR("Bad ML model handle\n");
+		result_local.error_code = ML_BAD_HDL;
+		goto load_fail;
+	}
+
+	odp_ticketlock_lock(&mdl->lock);
+	if (odp_unlikely(mdl->state == ML_STATE_FREE)) {
+		_ODP_ERR("Model has not been created yet or has been destroyed\n");
+		odp_ticketlock_unlock(&mdl->lock);
+		result_local.error_code = ML_NOT_CREATED;
+		goto load_fail;
+	}
+
+	if (odp_unlikely(mdl->num_loaded >= _odp_ml_glb->ml_config.max_models_loaded)) {
```

Comment:
This counter should be global, not per model. Single model can be loaded only once, after that it needs to be unloaded (before loading again).

Reply:
Since ML_MAX_MODELS_CREATED and ML_MAX_MODELS_LOADED are the same, and each created model can only be loaded once, num_loaded has no purpose. Removed.